### PR TITLE
[Feature] Support Mask2former in MMSeg 1.x

### PR DESCRIPTION
**Pull request by xiexinch**
## Motivation

Support Mask2former.

Ref: 
1. [Official repo](https://github.com/facebookresearch/Mask2Former) 
2. [MMDetection](https://github.com/open-mmlab/mmdetection)

## Modification

1. add mask2former head
2. configs

## Current results:

|dataset|backbone|mIoU|
|---|---|---|
|Cityscapes|R50|80.44|
|Cityscapes|R101|80.80|
|Cityscapes|swin-t|81.71|
|Cityscapes|swin-s|82.57|
|Cityscapes|swin-b (in22k)|83.52|
|Cityscapes|swin-l (in22k)|83.65|
|ADE20K|R50|47.87|
|ADE20K|R101|48.60|
|ADE20K|swin-t|48.66|
|ADE20K|swin-s|51.24|
|ADE20K|swin-b|52.44|
|ADE20K|swin-b (in22k)|53.90|
|ADE20K|swin-l (in22k)|56.01|

### CODE REVIEW

**Review by MeowZheng — COMMENTED** [review]
please add readme  metafile and ut

**Comment by MeowZheng** [comment]
update requirements/optional and add mmdet in it

File: mmseg/models/decode_heads/mask2former_head.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+            if len(masks) == 0:
+                gt_masks = torch.zeros(
+                    (0, gt_semantic_seg.shape[-2],
+                     gt_semantic_seg.shape[-1])).to(gt_semantic_seg).long()
+            else:
+                gt_masks = torch.stack(masks).squeeze(1).long()
+
+            instance_data = InstanceData(labels=gt_labels, masks=gt_masks)
+            batch_gt_instances.append(instance_data)
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/mask2former_head.py:63]:
```suggestion
    def _segdata2instancedata(self, batch_data_samples):
        batch_img_metas = []
        batch_gt_instances = []

        for data_sample in batch_data_samples:
            batch_img_metas.append(data_sample.metainfo)
            gt_semantic_seg = data_sample.gt_sem_seg.data
            classes = torch.unique(
                gt_semantic_seg,
                sorted=False,
                return_inverse=False,
                return_counts=False)

            # remove ignored region
            gt_labels = classes[classes != self.ignore_index]

            masks = []
            for class_id in gt_labels:
                masks.append(gt_semantic_seg == class_id)

            if len(masks) == 0:
                gt_masks = torch.zeros(
                    (0, gt_semantic_seg.shape[-2],
                     gt_semantic_seg.shape[-1])).to(gt_semantic_seg).long()
            else:
                gt_masks = torch.stack(masks).squeeze(1).long()

            instance_data = InstanceData(labels=gt_labels, masks=gt_masks)
            batch_gt_instances.append(instance_data)
    def loss(self, x: Tuple[Tensor], batch_data_samples: SampleList,
             train_cfg: ConfigType) -> dict:
        # batch_segdata_sample to batch_gt_instances
        batch_gt_instance = self._segdata2instancedata(batch_data_samples)
```

File: mmseg/models/decode_heads/mask2former_head.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            size = batch_img_metas[0]['pad_shape']
+        else:
+            size = batch_img_metas[0]['img_shape']
+        # upsample mask
+        mask_pred_results = F.interpolate(
+            mask_pred_results, size=size, mode='bilinear', align_corners=False)
+        cls_score = F.softmax(mask_cls_results, dim=-1)[..., :-1]
+        mask_pred = mask_pred_results.sigmoid()
+        sem_seg_mask = torch.einsum('bqc, bqhw->bchw', cls_score, mask_pred)
+        return sem_seg_mask
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/mask2former_head.py:100]:
```suggestion
        seg_logits = torch.einsum('bqc, bqhw->bchw', cls_score, mask_pred)
        return seg_logits
```

File: mmseg/models/decode_heads/mask2former_head.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+@MODELS.register_module()
+class Mask2FormerHead(MMDET_Mask2FormerHead):
+
+    def __init__(self,
+                 num_classes,
+                 align_corners=False,
+                 ignore_index=255,
+                 **kwargs):
+        super().__init__(**kwargs)
```
Comment (MengzhangLI) [inline mmseg/models/decode_heads/mask2former_head.py:30]: Add docstring: https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/decode_heads/maskformer_head.py#L23-L33

File: mmseg/models/decode_heads/mask2former_head.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        self.num_classes = num_classes
+        self.align_corners = align_corners
+        self.out_channels = num_classes
+        self.ignore_index = ignore_index
+
+        feat_channels = kwargs['feat_channels']
+        self.cls_embed = nn.Linear(feat_channels, self.num_classes + 1)
+
+    def _seg_data_to_instance_data(self, batch_data_samples: SampleList):
+
```
Comment (MengzhangLI) [inline mmseg/models/decode_heads/mask2former_head.py:41]: Add docstring followed [here](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/decode_heads/maskformer_head.py#L53-L71).

File: mmseg/models/decode_heads/mask2former_head.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                    (0, gt_sem_seg.shape[-2],
+                     gt_sem_seg.shape[-1])).to(gt_sem_seg).long()
+            else:
+                gt_masks = torch.stack(masks).squeeze(1).long()
+
+            instance_data = InstanceData(labels=gt_labels, masks=gt_masks)
+            batch_gt_instances.append(instance_data)
+        return batch_gt_instances, batch_img_metas
+
+    def loss(self, x: Tuple[Tensor], batch_data_samples: SampleList,
```
Comment (MengzhangLI) [inline mmseg/models/decode_heads/mask2former_head.py:72]: Add docstring followed [here](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/decode_heads/maskformer_head.py#L108-L121).

File: mmseg/models/decode_heads/mask2former_head.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        # forward
+        all_cls_scores, all_mask_preds = self(x, batch_data_samples)
+
+        # loss
+        losses = self.loss_by_feat(all_cls_scores, all_mask_preds,
+                                   batch_gt_instances, batch_img_metas)
+
+        return losses
+
+    def predict(self, x: Tuple[Tensor], batch_img_metas: List[dict],
```
Comment (MengzhangLI) [inline mmseg/models/decode_heads/mask2former_head.py:87]: Add docstring followed [here](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/decode_heads/maskformer_head.py#L137-L149).

**Review by MeowZheng — APPROVED** [review]

**Pull request merged**